### ISSUE TITLE
GZip decoder hangs in endless (pull?) loop

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#17417]
See https://groups.google.com/forum/#!topic/akka-user/a_FvKw3gX8E

/cc @sirthias

**Comment by drewhk** [comment]
can this be also related to the pushandfinish fix? since that was not yet released

**Comment by jrudolph** [comment]
It seems to happen on the current version as well.

Here's the stack trace:

```
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.removeBits(Interpreter.scala:259)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.removeBits(Interpreter.scala:450)
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.pull(Interpreter.scala:292)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.pull(Interpreter.scala:450)
	at akka.stream.stage.StageState.onPull(Stage.scala:299)
	at akka.stream.stage.StatefulStage.onPull(Stage.scala:362)
	at akka.stream.stage.StatefulStage.onPull(Stage.scala:323)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.run(Interpreter.scala:456)
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.progress(Interpreter.scala:235)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.progress(Interpreter.scala:450)
	at akka.stream.impl.fusing.OneBoundedInterpreter.akka$stream$impl$fusing$OneBoundedInterpreter$$execute(Interpreter.scala:563)
	at akka.stream.impl.fusing.OneBoundedInterpreter.akka$stream$impl$fusing$OneBoundedInterpreter$$unsafeFork(Interpreter.scala:627)
	at akka.stream.impl.fusing.OneBoundedInterpreter.akka$stream$impl$fusing$OneBoundedInterpreter$$fork(Interpreter.scala:615)
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.finish(Interpreter.scala:313)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.finish(Interpreter.scala:450)
	at akka.stream.impl.fusing.OneBoundedInterpreter$Finished$.onPull(Interpreter.scala:35)
	at akka.stream.impl.fusing.OneBoundedInterpreter$Finished$.onPull(Interpreter.scala:33)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.run(Interpreter.scala:456)
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.progress(Interpreter.scala:235)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.progress(Interpreter.scala:450)
	at akka.stream.impl.fusing.OneBoundedInterpreter.akka$stream$impl$fusing$OneBoundedInterpreter$$execute(Interpreter.scala:563)
	at akka.stream.impl.fusing.OneBoundedInterpreter.akka$stream$impl$fusing$OneBoundedInterpreter$$unsafeFork(Interpreter.scala:627)
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.pushAndFinish(Interpreter.scala:339)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.pushAndFinish(Interpreter.scala:450)
	at akka.stream.impl.fusing.BatchingActorInputBoundary.onPull(ActorInterpreter.scala:80)
	at akka.stream.impl.fusing.BatchingActorInputBoundary.onPull(ActorInterpreter.scala:24)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.run(Interpreter.scala:456)
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.progress(Interpreter.scala:235)
	at akka.stream.impl.fusing.OneBoundedInterpreter$$anon$3.progress(Interpreter.scala:450)
	at akka.stream.impl.fusing.OneBoundedInterpreter.akka$stream$impl$fusing$OneBoundedInterpreter$$execute(Interpreter.scala:563)
	at akka.stream.impl.fusing.OneBoundedInterpreter$State$class.execute(Interpreter.scala:231)
	at akka.stream.impl.fusing.OneBoundedInterpreter$EntryState.execute(Interpreter.scala:638)
	at akka.stream.stage.AbstractStage.enterAndPull(Stage.scala:72)
	at akka.stream.impl.fusing.ActorOutputBoundary.akka$stream$impl$fusing$ActorOutputBoundary$$tryPutBallIn(ActorInterpreter.scala:205)
	at akka.stream.impl.fusing.ActorOutputBoundary$$anonfun$downstreamRunning$1.applyOrElse(ActorInterpreter.scala:285)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:16)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:12)
	at scala.PartialFunction$class.applyOrElse(PartialFunction.scala:118)
	at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:12)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:165)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:166)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:467)
	at akka.stream.impl.fusing.ActorInterpreter.aroundReceive(ActorInterpreter.scala:339)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
	at akka.dispatch.Mailbox.run(Mailbox.scala:221)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by drewhk** [comment]
btw, if you turn on the Debug variable in the OneBoundedInterpreter companion object you can see tracing info (works best when used with isolated stages).

**Comment by jrudolph** [comment]
In the provided test case the compressor isn't flushed at the end, so in the end it seems to happens if you  feed truncated input to the decoder.

It may be related to this:

https://github.com/akka/akka/commit/e30c3e6492fdffb20ee8baa488611213b5a66755#diff-4b0c7b44fcdd24e6f5e04756cc50812aR32

**Comment by jrudolph** [comment]
You can add this test to `GzipSpec` for a reproducer:

```scala
    "throw an error if compressed data wasn't flushed properly at the end" in {
      def brokenCompress(payload: String) = Gzip.newCompressor.compress(ByteString(payload, "UTF-8"))

      val ex = the[ZipException] thrownBy ourDecode(brokenCompress("abcdefghijkl"))
      ex.getMessage should equal("Truncated GZIP stream")
    }
```

**Comment by drewhk** [comment]
Ok

**Comment by jrudolph** [comment]
I think it's quite simple:

```
output: PULL
input: PULL
input: ByteString(31, -117, 8, 0, 0, 0, 0, 0, 0, 0)
input: PULL
input: Complete
```

The `onUpstreamFinish` is then handled with `ctx.absorbTermination`. What could possibly go wrong?

**Comment by drewhk** [comment]
accidentally pulling while isFinishing? I don't know. Can you embed the stage in an IteratorInterpreter and turn on the Debug flag I mentioned?

**Comment by jrudolph** [comment]
I get this if I just run the test using a normal run with a simple pipeline (without using the Interpreter):

```
 enq ByteString(31, -117, 8, 0, 0, 0, 0, 0, 0, 0) akka://GzipSpec/user/$a/flow-1-1-stageFactory
 dem 1 (+1)              akka://GzipSpec/user/$a/flow-1-1-stageFactory
        ENTR             akka://GzipSpec/user/$a/flow-1-1-stageFactory
    <---                 akka://GzipSpec/user/$a/flow-1-1-stageFactory
<---                     akka://GzipSpec/user/$a/flow-1-1-stageFactory
---> ByteString(31, -117, 8, 0, 0, 0, 0, 0, 0, 0) akka://GzipSpec/user/$a/flow-1-1-stageFactory
<---                     akka://GzipSpec/user/$a/flow-1-1-stageFactory
---|                     akka://GzipSpec/user/$a/flow-1-1-stageFactory
Upstream was finished in state StartInflate
    <---                 akka://GzipSpec/user/$a/flow-1-1-stageFactory
```

**Comment by drewhk** [comment]
it looks like a pull after onUpstreamFinish, but I am not sure becuase the debug output is not reliable if there are potentially multiple interpreters running concurrently.

**Comment by jrudolph** [comment]
Yes, it's exactly that. In state `StartInflate` `onPull` isn't overridden to check if the stream has finished. I can fix it there but this also needs a fix for stages / in the interpreter to prevent pull while finishing.

**Comment by drewhk** [comment]
Please open a ticket, it should be easy to add.

**Comment by jrudolph** [comment]
Added #17418 to fix the endless loop and #17419 to fix the tests broken during e30c3e6 which seems to be another issue.

**Comment by jrudolph** [comment]
PR in #17420
